### DISCUSSION TITLE
Add Server Startup Hooks

### DISCUSSION THREAD

**Discussion by anthonyalayo** [vercel/next.js#43630] [Ideas]
### Describe the feature you'd like to request

This feature has been requested for over 2 years now in this discussion thread with no response from Vercel: https://github.com/vercel/next.js/discussions/15341

I belive @mschipperheyn [described it very simply:](https://github.com/vercel/next.js/discussions/15341#discussioncomment-3128585)

> Nextjs should have subscribeable events for this that accept promises:
> 
> ```
> preStartup
> postStartup
> preShutdown
> postShutdown
> ```
> The basic task of initialization is such a pain. You should not have to optin to a custom server to be able to do this.

This feature is **necessary** since standalone mode **does not work** with custom servers. Being able to initialize a prism connection pool or telemetry at the start of an application is standard for use with API routes.

### Describe the solution you'd like

Provide any sort of hook/function/subscribe-able event that developers can use to have code run once at server startup.

### Describe alternatives you've considered

I have considered using a custom server to do this, but it does not work with standalone mode. Standalone ignores your custom server and generates its own.

**Comment by anthonyalayo** [comment]
Please do not bucket this under discussion -- this is a feature request. The associated discussion is here: https://github.com/vercel/next.js/discussions/15341

**Comment by balazsorban44** [comment]
Hi, feature requests are converted to discussions as outlined in the contribution docs, as well as the issue template you filled out.

**Reply by jtlapp** [reply]
Wow this got punted without a response, neither here nor in the original discussion. Not a good look.

**Comment by puschie286** [comment]
Super basic feature - absolute necessary for every server application. Current solution with instrumentation is very unreliable.

I’m very surprised how nextjs could exist for almost 10 year without a “regular” server lifecycle.
